ledis/t_list.go

```diff
@@ -692,64 +693,39 @@ func (db *DB) LKeyExists(key []byte) (int64, error) {
 }
 
 func (db *DB) lblockPop(keys [][]byte, whereSeq int32, timeout time.Duration) ([]interface{}, error) {
-	ch := make(chan []byte)
-
-	bkeys := [][]byte{}
-	for _, key := range keys {
-		v, err := db.lpop(key, whereSeq)
-		if err != nil {
-			return nil, err
-		} else if v != nil {
-			return []interface{}{key, v}, nil
+	for {
+		var ctx context.Context
+		var cancel context.CancelFunc
+		if timeout > 0 {
+			ctx, cancel = context.WithTimeout(context.Background(), timeout)
 		} else {
-			db.lbkeys.wait(key, ch)
-			bkeys = append(bkeys, key)
-		}
-	}
-	if len(bkeys) == 0 {
-		return nil, nil
-	}
-
-	defer func() {
-		for _, key := range bkeys {
-			db.lbkeys.unwait(key, ch)
+			ctx, cancel = context.WithCancel(context.Background())
 		}
-	}()
 
-	deadT := time.Now().Add(timeout)
+		for _, key := range keys {
+			v, err := db.lbkeys.popOrWait(db, key, whereSeq, cancel)
 
-	for {
-		if timeout == 0 {
-			key := <-ch
-			if v, err := db.lpop(key, whereSeq); err != nil {
+			if err != nil {
```

Comment:
if we get the item or meet error, we should cancel the context when return to avoid the resource leak.